### DISCUSSION TITLE
20231101-optimizing-ci-pipeline/

### DISCUSSION THREAD

**Discussion by giscus** [inflearn/tech-blog-comment#45] [Comments]
# 20231101-optimizing-ci-pipeline/

Jenkins CI 환경에서 Turborepo 기반 모노레포 CI 소요시간을 최대 4.6배 개선한 경험과 pnpm 의존성 캐시 및 Turbo Task 캐시를 Jenkins 에이전트 간에 공유하는 방법을 소개합니다.

https://tech.inflab.com/20231101-optimizing-ci-pipeline/

**Comment by Sn-Kinos** [comment]
글 잘 읽었습니다! Yarn 과의 비교를 보던 중 궁금한 점이 생겨 질문 드립니다! pnpm 을 고려하셨다면 일반 Yarn 이 아닌 Yarn PnP 도 고려해보셨을 것으로 생각되는데 Yarn PnP 를 비교 대상에 넣지 않은 이유가 어떻게 되는 건가요?

**Reply by jakeinflab** [reply]
안녕하세요! 문의주신 내용에 답변드립니다.

yarn PnP 의 경우 지속적으로 Git 에 부하를 주고, 사용하지 않는 라이브러리를 삭제하더라도 Git에 영구적으로 남기 때문에 프로젝트가 커질 수록 부담이 될 수 있을 것이라고 판단했습니다.

제가 참고한 AB180 기술 블로그 내용을 공유드립니다.

> https://engineering.ab180.co/stories/yarn-to-pnpm
> 1/ Yarn PnP가 Git에 지속적으로 주는 부하
Airbridge는 Yarn Berry의 [PnP(Plug n' Play) 모드](https://yarnpkg.com/features/pnp)를 사용하여 패키지를 관리했습니다. node_modules 대신 .yarn/cache에 ZipFS 방식으로 패키지를 압축하여 저장하기에 패키지 설치 시간을 줄일 수 있기 때문입니다.
하지만 문득 PnP가 Git에 지속적으로 영향을 줄 수 있겠다는 생각이 들었습니다. 작업한 뒤 커밋을 하면 .git/objects 내에 변경된 파일이 zlib으로 압축되어 영구히 기록으로 남습니다.
> 고민은 Airbridge의 커밋 수는 10,000건이 넘는다는 점이었습니다. Airbridge에서 사용하는 패키지 중에서는 바이너리가 포함된 무거운 패키지들이 있습니다. 커밋의 수가 많을 수록, 한 커밋에 포함되는 파일의 용량과 수가 많을 수록, 이후의 작업에서 Git의 체크아웃, 브랜칭 속도에 영향을 줄 수 있겠다는 생각이 들었습니다.
이를 풀기 위해 Git 대용량 파일 저장소(LFS) 내지는 Git 최대 커밋 크기 조정, 주기적으로 [Git의 history를 살리면서 파일만 지우는 작업](https://stackoverflow.com/questions/8740187/git-how-to-remove-file-from-historical-commit)을 하는 것은 정말 오버 엔지니어링이라고 생각했습니다.

**Comment by sangbooom** [comment]
좋은 글 감사합니다~ 사용하신 [remote cache server](https://ducktors.github.io/turborepo-remote-cache/free-hosted-remote-cache.html)가 지원을 중단했는데 지금은 어떻게 대체해서 사용하고 계신지 궁금합ㄴ다!

**Reply by jakeinflab** [reply]
안녕하세요 sangboom님! 문의주신 내용에 답변드립니다.

전달해 주신 문서를 확인한 결과 `ducktors/turborepo-remote-cache` 에서 
미리보기 혹은 체험 목적으로 제공하던 **무료 리모트 캐시 서버의 운영을 중단**한 것으로 확인됩니다.

코드의 유지보수는 계속 되고 있고, 저희는 리모트 캐시 서버를 내부 사용 목적으로 별도 구축 및 운영하고 있습니다.

보안 측면에서도 제품 코드의 빌드 결과물을 외부에 보관하는 것은 가급적 피하는 것이 좋을 것 같습니다.

긴 글 읽어 주셔서 감사합니다!

**Reply by sangbooom** [reply]
많은 도움이 되었습니다! 답변 감사합니다 👍🏻
